compiler/rustc_hir_analysis/src/astconv/mod.rs

```diff
@@ -3137,16 +3139,36 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         debug!("impl_trait_ty_to_ty(def_id={:?}, lifetimes={:?})", def_id, lifetimes);
         let tcx = self.tcx();
 
-        let generics = tcx.generics_of(def_id);
+        // For new synthesized RPITITs the "parent" is the function that defined the RPITIT.
+        let generics_parent_count = match tcx.opt_rpitit_info(def_id) {
+            Some(ImplTraitInTraitData::Trait { fn_def_id, .. })
+            | Some(ImplTraitInTraitData::Impl { fn_def_id, .. }) => {
+                tcx.generics_of(fn_def_id).count()
+            }
+            _ => tcx.generics_of(def_id).parent_count,
+        };
 
-        debug!("impl_trait_ty_to_ty: generics={:?}", generics);
         let substs = InternalSubsts::for_item(tcx, def_id, |param, _| {
-            if let Some(i) = (param.index as usize).checked_sub(generics.parent_count) {
-                // Our own parameters are the resolved lifetimes.
-                let GenericParamDefKind::Lifetime { .. } = param.kind else { bug!() };
-                let hir::GenericArg::Lifetime(lifetime) = &lifetimes[i] else { bug!() };
-                self.ast_region_to_region(lifetime, None).into()
+            // When this `checked_sub` returns >= 0, it means that we are dealing with our own
+            // parameters and not the ones coming from parent.
+            if let Some(i) = (param.index as usize).checked_sub(generics_parent_count) {
```

Comment:
I don't think all of these changes are needed -- we just need to see if we're filling the last `lifetimes.len()` number of params:

```suggestion
            // We use `generics.count() - lifetimes.len()` here instead of `generics.parent_count`
            // since return-position impl trait in trait squashes all of the generics from its source fn
            // into its own generics, so the opaque's "own" params isn't always just lifetimes.
            if let Some(i) = (param.index as usize).checked_sub(generics.count() - lifetimes.len()) {
```

Then the rest of the changes are unnecessary.